Critical flow: Can see correct tax

alexflorisca opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR ads an e2e test to check that the right tax is being displayed, both in Cart and Checkout blocks. It also builds on the WC e2e-utils `shopper` functions and adds some useful ones related to blocks.

<!-- Reference any related issues or PRs here -->
Fixes #5772 and #5754 


<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->


### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

### Manual Testing

How to test the changes in this Pull Request:

1. `npm run wp-env`
2. Either run `npm run test:e2e -- Tax` 
3. Make sure all tests pass
4. Make sure e2e tests pass in CI

Commit:
Tax is displayed correctly e2e test

Commit:
Add extra utils to shopper

Commit:
add single e2e test script oin package.json

nielslange commented on bin/wp-env-config.sh:
```diff
@@ -18,5 +18,12 @@ wp rewrite flush
 wp core version --extra
 wp plugin list
 wp theme activate storefront
+wp plugin is-installed Basic-Auth
+if [ $? -eq 1 ]
+then
+  wp plugin install https://github.com/WP-API/Basic-Auth/archive/master.zip --activate
+else 
+  wp plugin activate Basic-Auth
+fi
```
After approving this PR before, I noticed the following section. This plugin should already be installed and activated by default due to the following entry:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/c8a5041817eb672a8490453731bd3dccc5dfa58a/.wp-env.json#L3-L7

Therefore, I'd say the following section is not necessary, isn't it?

alexflorisca replied:
Yep, good spot, I wasn't aware this even existed, thanks for pointing it out 👍

nielslange commented on package.json:
```diff
@@ -65,6 +65,7 @@
 		"storybook:deploy": "rimraf ./storybook/dist/* && npm run storybook:build && gh-pages -d ./storybook/dist",
 		"test": "wp-scripts test-unit-js --config tests/js/jest.config.json",
 		"test:e2e": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js",
+		"test:e2e-single": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js -t '<NAME OF YOUR TEST>' --watch --puppeteer-interactive",
```
I'm uncertain if we need this command, as we can easily run an individual test by specifying the test path. E.g. `npm run test:e2e-single Tax` could also be run with `npm run test:e2e-dev-watch -- tests/e2e/specs/shopper/tax.test.js`. 

I'd rather use the current commands and pipe the test URL into these commands, when required. Personally, I'd only introduce a new command if the desired action cannot be achieved with the existing commands. What are your thoughts on this?

alexflorisca replied:
Yeah you're right here, I didn't realise you could pass params to the underlying jest script through the current npm script. I'll remove this and use your way of doing it, much neater, thanks 👍

nielslange reviewed (commented):
Overall, your PR looks fantastic! 👋 I just added two minor questions to it.

Commit:
Feedback changes from Niels

alexflorisca commented:
Thanks for review @nielslange , I've made the change you requested. I also noticed that I haven't checked the order confirmation page as your issue describes, so I'll add that in and re-request the review in a bit

Commit:
Test tax is correct on summary page

Commit:
Merge branch 'trunk' into add/e2e-shopper-checkout-tax-test

Commit:
Check if CI test passes

Commit:
Merge branch 'trunk' into add/e2e-shopper-checkout-tax-test

Commit:
Longer timeout for tax test

Commit:
change timeout in jest config

Commit:
increate jest timeout

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
setTimeout outside of tests to 120000

ralucaStan commented on tests/e2e/specs/shopper/tax.test.js:
```diff
@@ -0,0 +1,80 @@
+	( prod ) => prod.name === 'Woo Single #1'
+);
+
+jest.setTimeout( 120000 );
```
why is this needed?

alexflorisca replied:
These tests are timing out, but only on CI. This is just me trying to increase the timeout to see if it fixes it (it didn't, it doesn't seem to increase the timeout at all). Weirdly this doesn't work inside a test or inside any of the `before` hooks which is why I've tried putting it out here

ralucaStan replied:
I had a test timing out for me locally. We need to increase the general setting, as we will have a lot of _longer_ tests. Can you try to only add this new value [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/2bfecbcf4ac26ea3af89c5929864ab119bd3263b/tests/e2e/config/jest.setup.js#L13-L14) and see the result.

alexflorisca replied:
Yep this worked, thanks :)

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -47,6 +58,148 @@ export const shopper = {
+	},
+
+	placeOrder: async () => {
+		await sleep( 1 );
```
Could [uiUnblocked](https://www.npmjs.com/package/@woocommerce/e2e-utils) be used instead?

alexflorisca replied:
That was my first try, I'm not sure how it works in woocommerce, when the `.blockUI` class is added on which the check is based, but it didn't work in this case. I suspect because we have to wait until the data store is updated, at which time there might be some UI changes and things can become unresponsive for some milliseconds. I know this isn't a neat solution. We could maybe try to write something more clever but I thought this was a good first pass. Happy to consider alternatives though

ralucaStan replied:
I've made this suggestion below based on the WC implementation. @nielslange  also added a similar utility

https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5860#discussion_r813802506

ralucaStan replied:
I've made this suggestion based on the WC implementation. @niels also added a similar utility

https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5860#discussion_r813802506

ralucaStan replied:
I've made this suggestion based on the WC implementation. @niels also added a similar utility

https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5860#discussion_r813802506

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -47,6 +58,148 @@ export const shopper = {
+		await page.click( '.wc-block-components-checkout-place-order-button' );
+	},
+
+	clearAndTypeInput: async ( selector, text ) => {
```
We don't need this if we use the `await expect(page).toFill('input[name="firstName"]', 'James')` from [expect-puppeteer](https://www.npmjs.com/package/expect-puppeteer), which automatically clears the input

ralucaStan commented on tests/e2e/specs/shopper/tax.test.js:
```diff
@@ -0,0 +1,80 @@
+		} );
+	} );
+} );
+jest.setTimeout( 60000 );
```
Same question, I wonder why this is needed.

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -2,15 +2,26 @@
+		await page.goto( checkoutBlockPermalink, {
+			waitUntil: 'networkidle0',
+		} );
+		await page.waitForSelector( 'h1', { text: 'Cart' } );
```
I know this is something present in other tests, but the API for waitForSelector does not allow to pass text here. https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pagewaitforselectorselector-options

Better to use `await expect(page).toMatchElement('h1', { text: 'Cart' })` from expect-pupetter.

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -47,6 +58,148 @@ export const shopper = {
+			couponCode
+		);
+		await page.click( '.wc-block-components-totals-coupon__button' );
+		await page.waitForSelector( '.wc-block-components-chip__text', {
```
again, this is incorrect, we need to use `await expect(page).toMatchElement('.wc-block-components-chip__text', { text: couponCode })`

ralucaStan replied:
This also does some waiting until the selector is in the page.

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -47,6 +58,148 @@ export const shopper = {
+		*/
+	searchForProduct: async ( productname ) => {
+		const searchFieldSelector = 'input.wp-block-search__input';
+		await page.waitForSelector( searchFieldSelector, {
```
`await expect(page).toMatchElement()` will wait for the element

ralucaStan reviewed (dismissed):
@alexflorisca Really nice work on testing taxes, nobody likes doing taxes :)

I've left some comments

Commit:
Address raluca's feedback and timeout everywheregaa

Commit:
Increase timeout on  for search box

Commit:
set jest timeout to 120000

Commit:
Raluca's suggestions

Commit:
Debug

senadir commented:
> I've added the [Basic-Auth](https://github.com/WP-API/Basic-Auth) plugin in order to be able to use the WC Rest API helper withRestApi from the [e2e-utils](https://www.npmjs.com/package/@woocommerce/e2e-utils) package to toggle WC settings.

This should already be added as we use it in our fixtures.

Commit:
Merge trunk

alexflorisca commented:
> > I've added the [Basic-Auth](https://github.com/WP-API/Basic-Auth) plugin in order to be able to use the WC Rest API helper withRestApi from the [e2e-utils](https://www.npmjs.com/package/@woocommerce/e2e-utils) package to toggle WC settings.
> 
> This should already be added as we use it in our fixtures.

Yeah @nielslange pointed this out as well, thanks @senadir , I've removed the extra invocation of the plugin

Commit:
Increase search timeout to 5000

Commit:
Only run Tax e2e tests in CI for quicker debug - REMOVE THIS

Commit:
use waitforSelector

Commit:
30s delay for toMatchElement

Commit:
Let'shope this works

Commit:
Remove toMatchElement

Commit:
waitForSelector

Commit:
Run debug test only

Commit:
Debug tax test

Commit:
debug taxes

Commit:
shopper using toMatchElement again

Commit:
Remove the 30000 timeout delay from toMatchElement

Commit:
Run all tests but only once

Commit:
Remove tax test

Commit:
Revert "Remove tax test"

This reverts commit 7db34120e1e91f4fd26514fdb9525cac2a6066bb.

Commit:
Delete the minicart test

Commit:
Revert "Revert "Remove tax test""

This reverts commit 31dd654e52f37fcd02ccd25336958248bef495c6.

Commit:
Revert "Delete the minicart test"

This reverts commit 5595f0834a975d097a300bc0c7cbf643caa8b764.

Commit:
tax test

Commit:
Merge trunk

Commit:
Merge branch 'trunk' into add/e2e-shopper-checkout-tax-test

Commit:
block addToCart

Commit:
Put back block functions

Commit:
Refactor block.addToCart

Commit:
mini-cart e2e test uses shopper.block.emptyCart()

Commit:
only run shopper tests in CI

Commit:
Run frontend and shopper tests and change waitFor -> waitForTimeout

Commit:
Run frontend tests in CI

Commit:
Run CI e2e tests again

Commit:
Skip legacy template blocks

Commit:
Skip legacy template test but run all others

Commit:
Switch to storefront instead of twentytwentyone after legacy-template-blocks tests finish

Commit:
Tidy up

Commit:
fix typo

alexflorisca commented on tests/e2e/specs/frontend/legacy-template-blocks.test.js:
```diff
@@ -43,7 +43,7 @@ describe( 'Legacy Template blocks', () => {
 
 	afterAll( async () => {
-		await activateTheme( 'twentytwentyone' );
+		await activateTheme( 'storefront' );
```
Switching theme to twentytwentyone makes the following tests fail, as most tests expect storefront as the theme. I will update this in a future PR to use the `useTheme` hook as @tomasztunik pointed out

nielslange reviewed (approved):
Well done, @alexflorisca! Let's 🚢 it! 🙌

tarhi-saad committed:
Merge branch 'trunk' into add/e2e-shopper-checkout-tax-test

Commit:
Fix ESLint issue
